### PR TITLE
Introduce swappable view model for chat and transcript

### PR DESCRIPTION
**Pull request by igorkravchenko**
Add ability to change view models for Chat view controller to reuse it in secure messaging.

MOB-1746

### CODE REVIEW

File: GliaWidgets/SecureConversations/Transcript/TranscriptCoordinator.swift [hunk 1]
```diff
@@ -0,0 +1,32 @@
+extension SecureConversations {
+    class TranscriptCoordinator: FlowCoordinator {
```
Comment (EgorovEI) [inline GliaWidgets/SecureConversations/Transcript/TranscriptCoordinator.swift:2]: final?

**Review by EgorovEI — APPROVED** [review]

**Comment by EgorovEI** [comment]
Awesome! Great job!

File: GliaWidgets/SecureConversations/Transcript/TranscriptCoordinator.swift (hunk 1 above)
Comment (igorkravchenko) [reply GliaWidgets/SecureConversations/Transcript/TranscriptCoordinator.swift]: Added a66194170874eaa11c7bd75757115d6510405861.

File: GliaWidgets/SecureConversations/Transcript/TranscriptCoordinator.swift [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        }
+
+        var delegate: ((DelegateEvent) -> Void)?
+        var environment: Environment
+
+        init(environment: Environment) {
+            self.environment = environment
+        }
+
+
```
[the remark this answers is not in the document]
Comment (igorkravchenko) [reply GliaWidgets/SecureConversations/Transcript/TranscriptCoordinator.swift]: Fixed fc95cb1daae11cd14479dee066d3b3b79656ae24.

File: GliaWidgets/SecureConversations/SecureConversations.Coordinator.swift [hunk 3]
```diff
@@ -112,6 +112,8 @@ extension SecureConversations {
                     controller?.presentSettingsAlert(
                         with: conf, cancelled: cancelled
                     )
+                case .transcriptRequested:
```
[the remark this answers is not in the document]
Comment (igorkravchenko) [reply GliaWidgets/SecureConversations/SecureConversations.Coordinator.swift:115]: Sure 64993e6aac4dc030595ab967d2874d8ef096e2bc.

File: GliaWidgets/SecureConversations/SecureConversations.ChatWithTranscriptModel.swift [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+extension SecureChatModel where Chat == ChatViewModel, Transcript == SecureConversations.TranscriptModel {
+    typealias Action = SecureChatModel<Chat.Action, Transcript.Action>
+    typealias ActionCallback = (Action) -> Void
+    typealias DelegateEvent = SecureChatModel<Chat.DelegateEvent, Transcript.DelegateEvent>
+    typealias DelegateCallback = (DelegateEvent) -> Void
+    typealias Event = SecureChatModel<Chat.Event, Transcript.Event>
+
+    var action: ActionCallback? {
```
[the remark this answers is not in the document]
Comment (igorkravchenko) [reply GliaWidgets/SecureConversations/SecureConversations.ChatWithTranscriptModel.swift:54]: Indeed. I have attempted to optimise  using some generic helpers. But still the problem is that for now Swift does not have something like KeyPath for enums. We could use PointFree's CasePath library, but again this is +1 dependency. We should decide on that as team.

File: GliaWidgets/SecureConversations/SecureConversations.ChatWithTranscriptModel.swift [hunk 5]
```diff
@@ -0,0 +1,220 @@
+enum SecureChatModel<Chat, Transcript> {
```
Comment (yurii-glia) [inline GliaWidgets/SecureConversations/SecureConversations.ChatWithTranscriptModel.swift:1]: To keep SPM supported we need to `import Foundation`, let's add it here.

**Review by yurii-glia — APPROVED** [review]

**Review by gersonnoboa — APPROVED** [review]

**Comment by igorkravchenko** [comment]
!squash

**Comment by igorkravchenko** [comment]
!merge

**Pull request merged**